### PR TITLE
Path subgroup access

### PR DESCRIPTION
**Pull request by AntoineRenaud91**
## Summary
This pull request introduces new methods for the `File`, `MutableFile`, `Group`, and `MutableGroup` structs. These methods enable users to fetch a specific subgroup, variable, or attribute within a netCDF structure using a relative path formatted like a Linux file system ("group1/group2/.../var").

## Key Features
- Enables direct querying of nested structures in netCDF files.
- Eliminates the need to keep intermediate subgroups alive when only interested in a specific, deeply-nested subgroup, variable, or attribute.

## Example Unit Test
Here's a unit test to demonstrate how the new feature works (can be found in netcdf/tests/file.rs):
```rust
#[test]
fn fetch_from_path() {
    // Create a temporary directory and netCDF file
    let d = tempfile::tempdir().unwrap();
    let path = d.path().join("cdf5.nc");
    {
        let mut file = netcdf::create(path.clone()).unwrap();
        let mut group = file.add_group("grp").unwrap();
        let mut subgroup = group.add_group("subgrp").unwrap();
        subgroup.add_dimension("dim", 1).unwrap();
        subgroup.add_variable::<f64>("var", &["dim"]).unwrap();
        subgroup.add_attribute("attr", "test").unwrap();
    }

    // Open the netCDF file and test the new methods
    let file = netcdf::open(path).unwrap();
    assert_eq!(
        file.group_from_path("grp/subgrp")
            .unwrap()
            .unwrap()
            .variable("var")
            .unwrap()
            .name(),
        file.variable_from_path("grp/subgrp/var").unwrap().name(),
    );

    match file
        .attribute_from_path("grp/subgrp/attr")
        .unwrap()
        .value()
        .unwrap()
    {
        netcdf::AttrValue::Str(string) => assert_eq!(string, "test"),
        _ => panic!(),
    }
}
```

This feature enhances user experience by providing a more intuitive and efficient way to navigate complex netCDF files.

### CODE REVIEW

**Comment by AntoineRenaud91** [comment]
I am not sure how to solve the last failed test. It seems that the test is not finding the *libnetcdf* library, which I don't believe is likely due to my changes.

**Comment by magnusuMET** [comment]
This feature looks great! In fact, why don't we make this the default so we could do
```rust
let var = file.variable("my/group/here/variable")?;
```
This works like the official python library and makes groups much more pleasant compared to the lifetime struggle with the old approach.

You can ignore the failing test and I will try to figure out what is going wrong there. It is not related to your changes

**Comment by AntoineRenaud91** [comment]
I'm glad you like this feature. Should I make a commit to set it as the default behavior and subsequently remove all the '_from_path' methods? It seems very unlikely that existing projects using this crate would have group or variable names with slashes.

**Comment by magnusuMET** [comment]
Would be great if you made that commit. Slashes are not allowed in object names in hdf5, so I would think they need to be disallowed in `netCDF` too.

**Comment by AntoineRenaud91** [comment]
Done ! :)

**Review by magnusuMET — COMMENTED** [review]
Just avoiding an unnecessary allocation.

I know the scope has increased a bit from the original idea, but this is a great ergonomic boost. Some other cases I have though up:
 - Dimensions should also allow a group name
 - What should we do if we try to add a variable `in/a/group`? Automatically create the group?
 - What should we return if we request a variable in a group which does not exist? Return `None` or an error?

File: netcdf/src/group.rs [hunk 1]
```diff
@@ -299,17 +299,20 @@ pub(crate) fn groups_at_ncid<'f>(ncid: nc_type) -> error::Result<impl Iterator<I
...
 
-pub(crate) fn group_from_name<'f>(ncid: nc_type, name: &str) -> error::Result<Option<Group<'f>>> {
-    let byte_name = super::utils::short_name_to_bytes(name)?;
-    let mut grpid = 0;
-    let e = unsafe {
-        super::with_lock(|| nc_inq_grp_ncid(ncid, byte_name.as_ptr().cast(), &mut grpid))
-    };
-    if e == NC_ENOGRP {
-        return Ok(None);
+pub(crate) fn group_from_path<'f>(ncid: nc_type, path: &str) -> error::Result<Option<Group<'f>>> {
```
[the remark this answers is not in the document]
Comment (AntoineRenaud91) [reply netcdf/src/group.rs:302]: Should we then update all the crate backend functions—`group_from_path`, `variable_from_path`, and `attribute_from_path`—to take an `impl Iterator<Item = &'i str>`as a path and perform the `.split("/")` operation within the bodies of the higher-level methods `group`, `variable`, and `attribute`?

[the remark this answers is not in the document]
Comment (magnusuMET) [reply netcdf/src/group.rs:302]: I am OK with just using `.split('/')` within the bodies since that is where we decide which group we are working on

**Comment by AntoineRenaud91** [comment]
> Just avoiding an unnecessary allocation.
> 
> I know the scope has increased a bit from the original idea, but this is a great ergonomic boost. Some other cases I have though up:
> 
>     * Dimensions should also allow a group name
> 
>     * What should we do if we try to add a variable `in/a/group`? Automatically create the group?
> 
>     * What should we return if we request a variable in a group which does not exist? Return `None` or an error?
I think the API would be safer if we allowed recursive group creation only through the add_group method. I would expect all other "add" methods to return an error if the subgroup structure doesn't exist.

As for requesting, I think it should return an error if the parent groups do not exist.

**Comment by magnusuMET** [comment]
We should probably add a new error variant, but that would be a breaking change which we could avoid by returning an `Error::Str` (and a TODO on `Error` for when we have to break the API)

**Comment by AntoineRenaud91** [comment]
I just commited these changes.  I created a function called 'crate' to transform an 'ncid' and a 'path' string into the 'ncid' of the deepest subgroup and the stem name. It returns an error if any of the subgroups do not exist. This way, I didn't have to make significant modifications to all the 'add' methods. I simply used this function to update the 'ncid' and the stem name.

File: netcdf/src/attribute.rs [hunk 2]
```diff
@@ -711,12 +711,16 @@ impl<'a> Attribute<'a> {
...
-
-    pub(crate) fn find_from_name(
-        ncid: nc_type,
+    pub(crate) fn find_from_path<'i>(
+        mut ncid: nc_type,
         varid: Option<nc_type>,
-        name: &str,
+        mut path: impl Iterator<Item = &'i str> + DoubleEndedIterator,
     ) -> error::Result<Option<Self>> {
+        let name = path.next_back().unwrap_or("");
```
Comment (magnusuMET) [inline netcdf/src/attribute.rs:719]: Use `get_subgroup_ncid_and_stem_from_path` here

File: netcdf/src/file.rs [hunk 3]
```diff
@@ -216,7 +215,12 @@ impl File {
     ///
     /// Not a `netCDF-4` file
     pub fn group<'f>(&'f self, name: &str) -> error::Result<Option<Group<'f>>> {
-        super::group::group_from_name(self.ncid(), name)
+        super::group::group_from_path(self.ncid(), name.split('/')).map(|ncid| {
```
Comment (magnusuMET) [inline netcdf/src/file.rs:218]: Use `get_subgroup_ncid_and_stem_from_path` here to only return `None` if the last group is missing

File: netcdf/src/group.rs [hunk 4]
```diff
@@ -299,19 +300,51 @@ pub(crate) fn groups_at_ncid<'f>(ncid: nc_type) -> error::Result<impl Iterator<I
...
-        _file: PhantomData,
-    }))
+pub(crate) fn get_subgroup_ncid_and_stem_from_path(
+    ncid: nc_type,
+    path: &str,
+) -> error::Result<(nc_type, &str)> {
+    let mut path = path.split('/');
+    let name = path.next_back().unwrap();
+    let ncid = super::group::group_from_path(ncid, path)?
+        .ok_or(error::Error::Str("Missing child group".into()))?;
```
Comment (magnusuMET) [inline netcdf/src/group.rs:348]: Use `ok_or_else` to avoid needlesly constructing a string

File: netcdf/src/variable.rs [hunk 5]
```diff
@@ -58,7 +58,15 @@ pub enum Endianness {
 
 #[allow(clippy::len_without_is_empty)]
 impl<'g> Variable<'g> {
-    pub(crate) fn find_from_name(ncid: nc_type, name: &str) -> error::Result<Option<Variable<'g>>> {
+    pub(crate) fn find_from_path<'i>(
+        mut ncid: nc_type,
+        mut path: impl Iterator<Item = &'i str> + DoubleEndedIterator,
+    ) -> error::Result<Option<Variable<'g>>> {
+        let name = path.next_back().unwrap_or("");
```
Comment (magnusuMET) [inline netcdf/src/variable.rs:65]: Use `get_subgroup_ncid_and_stem_from_path` here

**Review by magnusuMET — COMMENTED** [review]
I like `get_subgroup_ncid_and_stem_from_path`, just need to make sure we are using it everywhere and handling the case when a subgroup does not exist as an error

**Comment by magnusuMET** [comment]
Thanks a lot for your work on this @AntoineRenaud91, this is an awesome feature!

**Comment by AntoineRenaud91** [comment]
Happy to provide a useful contribution :)

**Pull request merged**